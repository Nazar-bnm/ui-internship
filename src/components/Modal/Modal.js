import React, { useEffect, useRef } from 'react';
import cx from 'classnames';

import './Modal.scss';

const CN = 'modal';

const Modal = (props) => {
  const {
    show,
    children,
    hideModal,
    className
  } = props;
  const modalRef = useRef(null);

  useEffect(() => {
    if (show) {
      document.body.classList.add('remove-scroll');
      modalRef.current.focus();
    }
  });

  const handleHiding = () => {
    document.body.classList.remove('remove-scroll');
    hideModal();
  };

  const hideModalOnBackdrop = (event) => {
    event.target.classList.contains(CN)
      && handleHiding();
  };

  const hideModalOnEscape = (event) => {
    const escapeKeyCode = 27;

    event.keyCode === escapeKeyCode && handleHiding();
  };

  return (
    <div
      tabIndex="-1"
      ref={modalRef}
      className={cx(`${CN}`, { [`${CN}--hide`]: !show })}
      onClick={hideModalOnBackdrop}
      onKeyDown={hideModalOnEscape}
    >
      <div className={cx(`${CN}__window`, className, { [`${CN}__window--animation`]: show })}>
        {children}
        <span
          className={`${CN}__close-button`}
          role="button"
          label="closer"
          tabIndex="0"
          onClick={handleHiding}
        >
          <i className={cx('close', 'icon')} />
        </span>
      </div>
    </div>
  );
};

export default Modal;
