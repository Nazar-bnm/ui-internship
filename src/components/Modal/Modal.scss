@import '../../scss/global_styles';

.modal {
  @include flex($justify: center, $align: center);
  position: fixed;
  z-index: 100;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background-color: rgba($color: $black, $alpha: 0.5);
  outline: none;

  &--transition {
    transition: transform 0.5s cubic-bezier(0.165, 0.84, 0.44, 1);
    transform: scale(0);
    background-color: inherit;
  }

  &__window {
    position: relative;
    width: 700px;
    height: 400px;
    padding: 0 3rem;
    background: $white;
    animation: blowUpModal 0.5s cubic-bezier(0.165, 0.84, 0.44, 1) forwards;
  }

  &__close-button,
  i.icon {
    transition: all linear 200ms;
  }

  &__close-button {
    @include flex($justify: center, $align: center);
    position: absolute;
    top: 1rem;
    right: 1rem;
    border: 1px solid $silver-chalice;
    width: 2rem;
    height: 2rem;

    &:hover {
      border-color: $dove-gray;

      i.icon {
        color: $dove-gray;
      }
    }
  }

  i.icon {
    margin: 0;
    font-size: 0.8rem;
    cursor: pointer;
    color: $silver-chalice;
  }

  @keyframes blowUpModal {
    0% {
      transform: scale(0);
    }
    100% {
      transform: scale(1);
    }
  }
}

.remove-scroll {
  overflow: hidden;
}
