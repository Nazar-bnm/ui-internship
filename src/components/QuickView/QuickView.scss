@import "../../scss/global_styles";

.quick-view {
  max-height: 60vh;
  max-width: 80vw;
  padding: 1rem 3.5rem 1.5rem;

  .heading {
    &__wrapper {
      margin-bottom: 1rem;
      height: 2rem;
    }

    &__title {
      font-size: 1.5rem;
    }
  }

  &-content {
    display: flex;
    justify-content: space-evenly;
    height: 90%;

    .product-order {
      display: block;
      width: 100%;
    }

    .product-image {
      height: 100%;

      &__container {
        flex-wrap: nowrap;
        margin-top: 8px;
        max-width: 40%;
        padding-right: 1rem;
      }

      &__big-image-container {
        width: fit-content;
        max-height: 70%;
      }

      &__small-images-container {
        max-height: 30%;
      }

      &__small-image {
        height: 100%;
      }
    }

    &__wrapper {
      height: 100%;
    }

    .product-order {
      &__heading {
        font-size: 1.5rem;
        margin-top: 0;
      }

      &__dropdowns-wrapper {
        height: 7rem;
      }

      .dropdown {
        height: 2rem;
      }
    }

    &__link {
      display: flex;
      justify-content: flex-end;
      padding-top: 8px;
      margin-left: auto;
      color: $orange;

      &:active,
      &:hover {
        color: $dove-gray;
      }
    }

    .button {
      cursor: pointer;
      width: 120px;
      color: $white;
      background-color: $black;

      &:hover {
        color: $black;
        background-color: $white;
      }
    }
  }
}

@include media-for(tablet) {
  .quick-view {
    &-content {
      display: flex;
      height: 90%;

      .heading__title {
        font-size: 5rem;
      }

      .product-order__heading {
        font-size: 1rem;
        margin-top: 0;
      }
    }
  }
}
