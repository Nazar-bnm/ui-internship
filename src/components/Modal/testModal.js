import React, { Component } from 'react';
import Modal from './Modal';

class TestModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: false
    };
  }

  showModal = () => {
    this.setState({
      show: true
    });
  };

  removeModal = () => {
    this.setState({
      show: false
    });
  };

  render() {
    const { show } = this.state;
    return (
      <>
        <button type="button" onClick={this.showModal}>
          show modal
        </button>
        {show && (
          <Modal removeModal={this.removeModal.bind(this)}>
            <p>Hello</p>
          </Modal>
        )}
      </>
    );
  }
}

export default TestModal;
