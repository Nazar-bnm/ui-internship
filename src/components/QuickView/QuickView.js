import React, { Component } from "react";
import { NavLink } from "react-router-dom";

import Modal from "../Modal";
import ProductImage from "../ProductImage";
import ProductOrder from "../ProductOrder";
import Heading from "../Heading";
import { defaultImages } from "../../constants";
import { productOrderParameters } from "../../config/ProductOrderMockups";

import "./QuickView.scss";

const CN = "quick-view";

class QuickView extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: false
    };

    this.removeModal = this.removeModal.bind(this);
  }

  showModal = () => {
    this.setState({
      show: true
    });
  };

  removeModal = () => {
    this.setState({
      show: false
    });
  };

  render() {
    const { show } = this.state;
    const {
      product: { id }
    } = this.props;

    return (
      <>
        <i className={`eye icon ${CN}__icon`} onClick={this.showModal} />
        {show && (
          <Modal className={CN} removeModal={this.removeModal}>
            <Heading title="Quick view" />
            <div className={`${CN}-content`}>
              <ProductImage images={defaultImages} />
              <div className={`${CN}-content__wrapper`}>
                <ProductOrder {...productOrderParameters} />
                <NavLink
                  to={`/product-page/${id}`}
                  className={`${CN}-content__link`}
                >
                  See more details
                </NavLink>
              </div>
            </div>
          </Modal>
        )}
      </>
    );
  }
}

export default QuickView;
