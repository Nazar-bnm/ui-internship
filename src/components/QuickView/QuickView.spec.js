import React from "react";
import { shallow } from "enzyme";

import QuickView from "./QuickView";
import Modal from "../Modal";

describe("<QuickView />", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<QuickView product={{ id: 10 }} />);
  });

  it("should match shapshot <QuickView />", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render Modal component when click on the icon", () => {
    expect(wrapper.find(Modal)).toHaveLength(0);
    wrapper.find("i").simulate("click");
    expect(wrapper.find(Modal)).toHaveLength(1);
  });

  it("should set state prop show to false when removeModal prop is executed", () => {
    wrapper.setState({show: true});
    wrapper.find(Modal).prop('removeModal')();
    expect(wrapper.state().show).toBeFalsy();
  })
});
