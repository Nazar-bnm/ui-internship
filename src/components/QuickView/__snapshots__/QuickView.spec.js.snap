// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuickView /> should match shapshot <QuickView /> 1`] = `
<Fragment>
  <i
    className="eye icon quick-view__icon"
    onClick={[Function]}
  />
</Fragment>
`;
